syntax = "proto3";
import "player.proto";
import "item.proto";
import "location.proto";
import "google/protobuf/timestamp.proto";


message PBInteraction 
{
    repeated PBInteractionType types                = 1;
    int32 database_id                               = 2;
    string interaction_list_name                    = 3;
    string interaction_name                         = 4;
    string description                              = 5;
    int32 pos_x                                     = 6;
    int32 pos_y                                     = 7;
    repeated string map_icon_paths                  = 8;

    PBPostInteraction post_info                     = 9;
    PBZoneAccessInteraction zone_access_info        = 10;
    PBShopInteraction shop_info                     = 11;
    PBQuestInteraction quest_info                   = 12;
    PBDialogInteraction dialog_info                 = 13;
    PBCustomInteraction custom_info                 = 14;
}

message PBPostInteraction 
{
    repeated PBItemData resources               = 1;
    PBPlayerSkills post_skill                   = 2;
    int32 interact_xp                           = 3;
    bool needs_database                         = 4;
    repeated PBUpgrades upgrades                = 5;

    //Database
    int32 capacity_upgrade                      = 6;
    int32 gen_second_upgrade                    = 7;
    int32 fortune_upgrade                       = 8;
    int32 resource_stored                       = 9;
    google.protobuf.Timestamp last_collected    = 10;

}

message PBUpgrades 
{
    PBUpgradeType upgrade_type              = 1;
    repeated PBUpgradeInfo info             = 2;
}

message PBUpgradeInfo 
{
    float current_stat                      = 3;
    repeated PBItemData upgrade_req         = 4;
}

message PBZoneAccessInteraction 
{
    PBLocationID next_loc                   = 1;
    bool needs_database                     = 2;
    repeated PBUpgradeInfo unlock_info      = 3;

    //database
    int32 unlock_level                      = 4;
}

message PBShopInteraction
{

}

message PBDialogInteraction
{
    bool needs_database         = 1;
    repeated PBDialogInfo info  = 2;

    //database
    int32 current_dialog        = 3;
}

message PBDialogInfo 
{
    string text                                 = 1;
    string npc_image                            = 2;
    bool more_text                              = 3;
    repeated PBResponseOption response_options  = 4;
}

message PBResponseOption 
{
    string text         = 1;
    int32 next_text     = 2;
}

message PBQuestInteraction
{

}

message PBCustomInteraction 
{

}


enum PBInteractionType 
{
    NONE                                    = 0;
    POST                                    = 1;
    ZONE_ACCESS                             = 2;
    QUEST                                   = 3;
    SHOP                                    = 4;
    DIALOG                                  = 5;
    CUSTOM                                  = 6;
}

enum PBUpgradeType 
{
    CAPACITY                                = 0;
    GEN_SECOND                              = 1;
    FORTUNE                                 = 2;
}